marshal/marshal.go

```diff
@@ -28,91 +30,371 @@ const (
 		return nil, err
 	}
 
 	// Get offset coordinator so we can look up (and save) committed offsets later.
-	coordinatorConf := kafka.NewOffsetCoordinatorConf(groupID)
-	coordinator, err := broker.OffsetCoordinator(coordinatorConf)
+	coordinator, err := cluster.getOffsetCoordinator(groupID)
 	if err != nil {
 		return nil, err
 	}
 
 	m := &Marshaler{
 		quit:       new(int32),
-		rsteps:     new(int32),
+		cluster:    cluster,
 		instanceID: newInstanceID(),
 		clientID:   clientID,
 		groupID:    groupID,
-		kafka:      broker,
 		offsets:    coordinator,
-		producer:   broker.Producer(kafka.NewProducerConf()),
-		topics:     make(map[string]int),
-		groups:     make(map[string]map[string]*topicState),
-		jitters:    make(chan time.Duration, 100),
 	}
 
-	// Do an initial metadata fetch, this will block a bit
-	err = m.refreshMetadata()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to get metadata: %s", err)
+	return m, nil
+}
+
+// newInstanceID creates a new random instance ID for use inside Marshal messages. This
+// is generated new every time we restart.
+func newInstanceID() string {
+	// A UUID4 starts with 8 random characters, so let's use that as our instance ID.
+	// This should be a good tradeoff between randomness and brevity.
+	return uuid.New()[0:8]
+}
+
+// addNewConsumer is called when a new Consumer is created. This allows Marshal to keep
+// track of the consumers that exist so we can operate on them later if needed.
+func (m *Marshaler) addNewConsumer(c *Consumer) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.consumers = append(m.consumers, c)
+}
+
+// getClaimedPartitionState returns a topicState iff it is claimed by the current Marshaler.
+// Else, an error is returned. This is on the Marshaler becomes it's a helper to only return
+// a claim that is presently valid and owned by us.
+func (m *Marshaler) getClaimedPartitionState(topicName string, partID int) (
+	*topicState, error) {
+
+	// Get partition state of whatever happens to be here
+	topic := m.cluster.getPartitionState(m.groupID, topicName, partID)
+
+	topic.lock.RLock()
+	defer topic.lock.RUnlock()
+
+	if !topic.partitions[partID].isClaimed(m.cluster.ts) {
+		return nil, fmt.Errorf("Partition %s:%d is not claimed!", topicName, partID)
+	}
+
+	// And if it's not claimed by us...
+	if topic.partitions[partID].GroupID != m.groupID ||
+		topic.partitions[partID].ClientID != m.clientID {
+		return nil, fmt.Errorf("Partition %s:%d is not claimed by us!", topicName, partID)
+	}
+
+	return topic, nil
+}
+
+// Topics returns the list of known topics.
+func (m *Marshaler) Topics() []string {
+	return m.cluster.getTopics()
+}
+
+// Partitions returns the count of how many partitions are in a given topic. Returns 0 if a
+// topic is unknown.
+func (m *Marshaler) Partitions(topicName string) int {
+	return m.cluster.getTopicPartitions(topicName)
+}
+
+// Terminate is called when we're done with the marshaler and want to shut down.
+func (m *Marshaler) Terminate() {
```

Comment:
In order to actually shut down clusters, I think clusters need to track their Marshalers just like how the Marshalers track their consumers. If a cluster has no associated Marshalers, it should terminate. Otherwise I don't see how a cluster ever gets terminated (aside from a Kafka availability problem).